Bug 1188934 - Harden LockScreen against Settings race conditions

cr opened the pull request:
This adds an internal check to LockScreen.unlock() that detects whether
or not the settings observer callback has written the user-configured
settings value for passCodeEnabled. If it hasn't, unlock() gracefully
refuses to unlock, and logs an error message to the console.

It also introduces LockScreen.forceUnlock() that does not check for
the race condition (identical to the old .unlock() behavior). Please
use this if you really mean to unconditionally unlock and circumvent
the pass code check.

Commit:
Harden LockScreen against Settings race conditions

This adds an internal check to LockScreen.unlock() that detects whether
or not the settings observer callback has written the user-configured
settings value for passCodeEnabled. If it hasn't, unlock() gracefully
refuses to unlock, and logs an error message to the console.

It also introduces LockScreen.forceUnlock() that does not check for
the race condition (identical to the old .unlock() behavior). Please
use this if you really mean to unconditionally unlock and circumvent
the pass code check.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bbe91b1ed6c14ac1947d22294e528e19750fd79b

Commit:
Simplify race condition hardening

This hardens LockScreen against changing the unsafe default of
.passCodeEnabled: false to .passCodeEnabled: true. This has the effect
that a premature 'lockscreen-notification-request-activate-unlock' event
does not blindly unlock the device before the observer callback sets the
unsafe value in case of no pass code protection.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7811645032ef0b85de875df8770bd7978e1e85f9
